handle changing managed domains in hiveConfig

joelddiaz opened the pull request:
The hive operator would not relaunch the hive-controllers deployment when the only change to the HiveConfig was adding/removing an existing managed domain entry (leaving at least one managed domain).

Update the logic to create a new configmap with a new name (so that the controllers cannot see the old configmap contents on controller restart), and detect when changes are needed so that the hive-controllers deployment is re-deployed as necessary.

A notable change is that a HiveConfig with no managed domain entries will now have a configmap with an empty list of managed domains used in the hive-controllers deployment (the dns controllers already handle the case of an empty list of managed domains). Previously the hive-operator would simply not create the configmap at all.

staebler commented on pkg/operator/hive/hive.go:
```diff
@@ -86,9 +86,7 @@ func (r *ReconcileHiveConfig) deployHive(hLog log.FieldLogger, h *resource.Helpe
-	if len(instance.Spec.ManagedDomains) > 0 {
-		addManagedDomainsVolume(&hiveDeployment.Spec.Template.Spec)
-	}
+	addManagedDomainsVolume(&hiveDeployment.Spec.Template.Spec, mdConfigMap.Name)
```
Is there a technical reason why we had to change this behavior? It seems strange to me to have an empty configmap.

joelddiaz replied:
i though it made the code easier to always assume there will be a configmap. i originally was preserving the behavior of not using a configmap in the absence of any managed domains, and just was dealing with a lot of special case if-no-managed-domains code. I can restore the old behavior if the empty configmap is really that surprising.

staebler replied:
OK. If it trickles down to a lot of places where you need to deal with the special case, then there is not too much harm in having an empty configmap.

staebler commented on pkg/operator/hive/manageddns.go:
```diff
@@ -37,38 +42,105 @@ func (r *ReconcileHiveConfig) teardownLegacyExternalDNS(hLog log.FieldLogger) er
+
+// configureManagedDomains will create a new configmap holding the managed domains settings (if necessary), or simply
+// return the current configmap of the current deployment if the settings it contains match the desired settings.
+// It will also return the old configmap if a new one was created (so that it can be garbage collected).
```
The function never returns the old configmap. It returns either the new one or the existing retained one.

joelddiaz replied:
yep, that comment was for a previous version of behavior. will fix.

staebler commented on pkg/operator/hive/manageddns.go:
```diff
@@ -37,38 +42,105 @@ func (r *ReconcileHiveConfig) teardownLegacyExternalDNS(hLog log.FieldLogger) er
+	for i, cm := range configMapList.Items {
+		if reflect.DeepEqual(cmData, cm.Data) {
+			currentConfigMap = &configMapList.Items[i]
+			continue
```
This continue is unnecessary.

staebler commented on pkg/operator/hive/manageddns.go:
```diff
@@ -37,38 +42,105 @@ func (r *ReconcileHiveConfig) teardownLegacyExternalDNS(hLog log.FieldLogger) er
+		return nil, errors.Wrap(err, "failed to list config maps for managed domains")
+	}
+
+	for i, cm := range configMapList.Items {
```
This leaves multiple configmaps if there are multiple that have the correct, current data. Coming out of this function, there should only be a single configmap.

joelddiaz replied:
i didn't see a way that the operator could end up with multiple configmaps with correct, current data (unless someone comes in external to the code and starts making these configmaps) (or maybe the operator has multiple replicas running)

staebler replied:
I agree that it is unlikely. However, you are already looping through all of the configmaps so I figured it wouldn't be too hard to also ensure that we don't have duplicates. It is fine as is.

joelddiaz replied:
I changed the flow to catch this case.

staebler commented on pkg/operator/hive/manageddns.go:
```diff
@@ -37,38 +42,105 @@ func (r *ReconcileHiveConfig) teardownLegacyExternalDNS(hLog log.FieldLogger) er
+	configMapList := &corev1.ConfigMapList{}
+	labelSelector := map[string]string{configMapLabel: "true"}
+
+	currentConfigMap := &corev1.ConfigMap{}
```
Move this declaration after the call to `List` since it is not used until then.

staebler replied:
When there is not a good current configmap, return nil instead of an empty configmap.

staebler commented on pkg/operator/hive/manageddns.go:
```diff
@@ -37,38 +42,105 @@ func (r *ReconcileHiveConfig) teardownLegacyExternalDNS(hLog log.FieldLogger) er
+		return nil, err
+	}
+
+	if !reflect.DeepEqual(newConfigMapData, currentConfigMap.Data) {
```
We have already done this deep equal check in `getCurrentConfigMap`. There is no need to do it again. If you return nil from `getCurrentConfigMap` when there is no good configmap, then you can check for nil here to determine whether you need to create a new configmap.

staebler commented on pkg/operator/hive/manageddns.go:
```diff
@@ -37,38 +42,105 @@ func (r *ReconcileHiveConfig) teardownLegacyExternalDNS(hLog log.FieldLogger) er
 
-func addManagedDomainsVolume(podSpec *corev1.PodSpec) {
+func addManagedDomainsVolume(podSpec *corev1.PodSpec, configMapName string) {
+	mountPath := "/data/config"
```
```
const mountPath = "/data/config"
```

joelddiaz commented:
@staebler i think i've addressed all the PR feedback (minus the bit about not using a configmap at all in the absense of any managed domains in hiveconfig)

joelddiaz commented:
/test e2e

joelddiaz commented:
/test e2e

dgoodwin commented:
Needs a squash, removal of WIP tag, and passing e2e.

joelddiaz commented:
/test e2e

joelddiaz commented:
/test e2e

joelddiaz commented:
/test e2e

joelddiaz commented:
/test e2e

joelddiaz commented:
/test e2e

joelddiaz commented:
/test e2e

staebler commented:
The e2e test failures are legitimate. Retesting will not help.

```
Creating cluster deployment
time="2020-01-30T19:51:42Z" level=error msg="running the apply command failed" command=create-cluster error="error when creating \"/tmp/apply-974266781\": admission webhook \"clusterdeploymentvalidators.admission.hive.openshift.io\" denied the request: The base domain must be a child of one of the managed domains for ClusterDeployments with manageDNS set to true" stderr= stdout=
```

staebler commented on contrib/pkg/adm/managedns/enable.go:
```diff
@@ -191,15 +206,59 @@ func (o *Options) Run(dynClient client.Client, args []string) error {
+
+	log.Info("waiting for new hiveadmission deployment")
+	for i := 0; i < 24; i++ {
+		err := dynClient.Get(context.TODO(), key, deployment)
```
Use an new empty deployment each loop through. A `Get` does not clear out fields.

staebler replied:
It would be nice if we got in the habit of using a watch for things like this instead of looping. See https://github.com/kubernetes/client-go/blob/1a26190bd76a9017e289958b9fba936430aa3704/tools/watch/until.go#L129.

staebler commented on contrib/pkg/adm/managedns/enable.go:
```diff
@@ -191,15 +206,59 @@ func (o *Options) Run(dynClient client.Client, args []string) error {
+		err := dynClient.Get(context.TODO(), key, deployment)
+		if err != nil {
+			log.WithError(err).Fatal("error getting hiveadmission deployment")
+			return "", err
```
Is there a race condition where this may run prior to the operator creating the hiveadmission deployment for the first time? If that does happen, then the test will fail here since the `Get` will return a NotFound error.

joelddiaz replied:
Yes! I'll change it to catch the IsNotFound case and keep waiting for a new deployment to appear.

staebler commented on contrib/pkg/adm/managedns/enable.go:
```diff
@@ -165,7 +174,13 @@ func (o *Options) Run(dynClient client.Client, args []string) error {
-	// they all have the expected env var set.
+	// they all reference the new configmap.
+	log.Debug("getting new configmap with updated dns data")
+	dnsConfigMapName, err := getUpdatedManagedDNSConfig(dynClient, currHiveadmissionGeneration)
```
The hiveutil command will return an error if the managed DNS is unchanged because it will time out waiting for an update to the generation of the hiveadmission deployment.

staebler commented on contrib/pkg/adm/managedns/enable.go:
```diff
@@ -191,15 +206,59 @@ func (o *Options) Run(dynClient client.Client, args []string) error {
+func podUsingConfigMap(p corev1.Pod, cmName string) bool {
+	for _, vol := range p.Spec.Volumes {
+		if vol.Name == constants.ManagedDomainsVolumeName {
+			if vol.ConfigMap.Name == cmName {
```
Can this be `return vol.ConfigMap.Name == cmName`? Is there an expectation that there could be multiple volumes with the name of `ManagedDomainsVolumeName`?

joelddiaz replied:
I would hope not. I'll fix it up to just do the equality return immediately.

joelddiaz commented:
@staebler i made a bunch of changes so that we can store status into the HiveConfig object. I haven't squashed, so it's all in the most recent commit on this PR. I think your proposed approach makes things easier to reason about, PTAL.

staebler commented on contrib/pkg/adm/managedns/enable.go:
```diff
@@ -165,18 +168,39 @@ func (o *Options) Run(dynClient client.Client, args []string) error {
+		log.WithError(err).Fatal("failed to get new configmap name")
+	}
+
+	if err := waitForHiveAdmissionPods(dynClient, dnsConfigMapName); err != nil {
```
Do we need to do this advanced wait any more? Can we simply wait for the hiveadmission deployment to roll out?

staebler commented on pkg/apis/hive/v1/hiveconfig_types.go:
```diff
@@ -57,6 +57,13 @@ type HiveConfigStatus struct {
+
+	// ConfigApplied will be set by the hive operator to indicate whether or not the LastGenerationObserved
+	// was successfully reconciled.
+	ConfigApplied bool `json:"configApplied,omitempty"`
```
It would be nice if this was a condition that stipulated why operator failed to apply the config. But that can be written up as a separate Jira card.

dgoodwin commented:
I have pointed this at master instead of v1 branch.

joelddiaz commented:
@staebler some major changes with these most recent commits.

Primary change is using some of functionality from what powers 'kubectl rollout status --wait' (and all the changes that required).
Also addressed the renaming of the status field to 'observedGeneration'.
Now rebased on the master branch as well.

This will need some squashing, but I'm keeping them separate for now to hopefully make clear the most recent changes.

staebler commented on Gopkg.lock:
```diff
@@ -2301,6 +2301,7 @@
     "k8s.io/kubernetes/pkg/apis/core/validation",
     "k8s.io/kubernetes/pkg/capabilities",
     "k8s.io/kubernetes/pkg/client/metrics/prometheus",
+    "k8s.io/kubernetes/pkg/kubectl",
```
How is there a change to `Gopkg.lock` without any change to vendor files?

staebler replied:
I do see that those files are already in vendor.

joelddiaz replied:
i had originally vendored some additional stuff from k8s.io/kubernetes/pkg/kubectl, but ended up not using it, so when i re-vendored, this is what was left over. i'm quite confused, but each current run of 'dep ensure' keeps that change in there.

staebler commented on contrib/pkg/adm/managedns/enable.go:
```diff
@@ -160,112 +171,159 @@ func (o *Options) Run(dynClient client.Client, args []string) error {
+		},
+	}
+
+	preconditionFunc := func(store cache.Store) (bool, error) {
```
I don't think we need this precondition for our use case. We **want** to wait for the timeout if the deployment does not exist already.

joelddiaz replied:
maybe i'm missing something here. why would a deployment not exist? this is the hiveadmission pods we're talking about which should always be running on a healthy instance of hive, no?

staebler replied:
This hiveutil command could be run prior to the hive-operator creating the hiveadmission deployments for the first time.

staebler commented on contrib/pkg/adm/managedns/enable.go:
```diff
@@ -160,112 +171,159 @@ func (o *Options) Run(dynClient client.Client, args []string) error {
-	for i, vol := range deployment.Spec.Template.Spec.Volumes {
-		if vol.Name == constants.ManagedDomainsVolumeName {
-			return deployment.Spec.Template.Spec.Volumes[i].ConfigMap.Name, nil
+	preconditionFunc := func(store cache.Store) (bool, error) {
```
We do no need this precondition.

staebler commented on contrib/pkg/adm/managedns/enable.go:
```diff
@@ -160,112 +171,159 @@ func (o *Options) Run(dynClient client.Client, args []string) error {
+
+	fieldSelector := fields.OneTermEqualSelector("metadata.name", hiveAdmissionDeployment).String()
+
+	lw := &cache.ListWatch{
```
Use `NewFilteredListWatchFromClient` from k8s.io/client-go/tools/cache.

staebler commented on contrib/pkg/adm/managedns/enable.go:
```diff
@@ -160,112 +171,159 @@ func (o *Options) Run(dynClient client.Client, args []string) error {
-		return "", err
+	fieldSelector := fields.OneTermEqualSelector("metadata.name", hiveConfigName).String()
+
+	lw := &cache.ListWatch{
```
Use NewFilteredListWatchFromClient from k8s.io/client-go/tools/cache.

staebler commented on contrib/pkg/adm/managedns/enable.go:
```diff
@@ -160,112 +171,159 @@ func (o *Options) Run(dynClient client.Client, args []string) error {
+	_, err = clientwatch.UntilWithSync(ctx, lw, &unstructured.Unstructured{}, preconditionFunc, func(e watch.Event) (bool, error) {
+		switch t := e.Type; t {
+		case watch.Added, watch.Modified:
+			_, done, err := statusViewer.Status(e.Object.(runtime.Unstructured), deployment.Generation)
```
I would pass in a revision of 0 instead of the revision from the deployment. We do not care about reaching that particular revision: We only care about reaching a point where the observedGeneration and generation are equal.

This also means that we do not need to get the deployment prior to the call to `UntilWithSync`.

staebler commented on contrib/pkg/adm/managedns/enable.go:
```diff
@@ -160,112 +171,159 @@ func (o *Options) Run(dynClient client.Client, args []string) error {
-	key := types.NamespacedName{Name: hiveAdmissionDeployment, Namespace: hiveNamespace}
-	deployment := &appsv1.Deployment{}
+func waitForHiveConfigToBeProcessed(hiveClient *hiveclient.Clientset) error {
+	// Wait for the HiveConfig.Status.ObservedGeneration to match
```
Put this comment on the function itself rather than inside the function.

staebler commented on contrib/pkg/adm/managedns/enable.go:
```diff
@@ -320,3 +378,30 @@ func (o *Options) getResourceHelper() (*resource.Helper, error) {
+	log.Debug("creating cluster client config")
+	rules := clientcmd.NewDefaultClientConfigLoadingRules()
+	kubeconfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(rules, &clientcmd.ConfigOverrides{})
+	cfg, err := kubeconfig.ClientConfig()
```
`GetClientConfig` from contrib/pkg/utils

staebler commented on contrib/pkg/adm/managedns/enable.go:
```diff
@@ -157,47 +171,159 @@ func (o *Options) Run(dynClient client.Client, args []string) error {
+	_, err = clientwatch.UntilWithSync(ctx, lw, &unstructured.Unstructured{}, preconditionFunc, func(e watch.Event) (bool, error) {
+		switch t := e.Type; t {
+		case watch.Added, watch.Modified:
+			_, done, err := statusViewer.Status(e.Object.(runtime.Unstructured), deployment.Generation)
```
This conversion from a `runtime.Object` to a `runtime.Unstructured` should have an OK check.

staebler commented on contrib/pkg/adm/managedns/enable.go:
```diff
@@ -381,12 +343,9 @@ func (o *Options) getResourceHelper() (*resource.Helper, error) {
+	cfg, err := hiveutils.GetClientConfig()
 	if err != nil {
 		log.WithError(err).Error("cannot obtain client config")
-		return err
```
Was the remove to the return intentional?

joelddiaz replied:
absolutely not. good catch.
